### PR TITLE
[core] Get data from nodes table

### LINKED ISSUE #8516: [KW-Search] Update `tables` GET  endpoints
Modify existing table GET endpoint to include data from nodes tables, and read parents from there

### Related design doc section:
[Changes in tables endpoint](https://www.notion.so/dust-tt/Design-doc-keyword-search-infrastructure-WIP-12228599d94180a6b19aeec0b493cb13?pvs=4)

### Todo:
- Update GET endpoints to join with nodes table
- Update documents/search endpoint to join with nodes table

### Depends on task(s):
#8518

### PR DESCRIPTION
**Pull request by tdraier**
fixes: https://github.com/dust-tt/dust/issues/8516 
fixes: https://github.com/dust-tt/dust/issues/8519

## Description

Join with nodes tables to fetch mimetype and title when querying documents or tables.
If entries are missing in nodes tables, fallback to the default values.

## Risk

Issues on document/table read, can be rolled back. 
This includes listing / searching nodes in data source, but also query-less retrieval action ( "include" )

## Deploy Plan

deploy core

### CODE REVIEW

File: core/src/stores/postgres.rs
Comment (spolu) [inline core/src/stores/postgres.rs:84]: let's type more and call it column with type Option<&str>

Right now a `.` is likely missing

Comment (philipperolet) [reply core/src/stores/postgres.rs:84]: Indeed
We could even go all the way for the type? e.g.
```
#[derive(Debug)]
enum TablePrefix {
    DataSourceDocument,  // "dsd."
    Table,              // "t."
    None,               // ""
}

impl TablePrefix {
    fn as_str(&self) -> &'static str {
        match self {
            TablePrefix::DataSourceDocument => "dsd.",
            TablePrefix::Table => "t.",
            TablePrefix::None => "",
        }
    }
}
```
and type column as TablePrefix ?

**Review by philipperolet — COMMENTED** [review]
LGTM modulo the typing topic, thanks :pray: 
Also, the cards marked fixed indicated that `parents[]` were to be read in data_source_nodes by the PR so this is why the backfills were required here
We can move forward but can you keep in mind to update the code once the backfills are done?

File: core/src/stores/postgres.rs
Comment (tdraier) [reply core/src/stores/postgres.rs:84]: 🙈  thanks for catching that

**Comment by tdraier** [comment]
> LGTM modulo the typing topic, thanks 🙏 Also, the cards marked fixed indicated that `parents[]` were to be read in data_source_nodes by the PR so this is why the backfills were required here We can move forward but can you keep in mind to update the code once the backfills are done?

yes I was planning to do that after the backfill - but this can be done here if we wait for the backfill to be done before  merging , as originally planned

**Comment by tdraier** [comment]
> > LGTM modulo the typing topic, thanks 🙏 Also, the cards marked fixed indicated that `parents[]` were to be read in data_source_nodes by the PR so this is why the backfills were required here We can move forward but can you keep in mind to update the code once the backfills are done?
> 
> yes I was planning to do that after the backfill - but this can be done here if we wait for the backfill to be done before merging , as originally planned

I updated to read the parents from nodes table. All documents and tables must be backfilled and have an entry in nodes. Now using inner join instead of outer join, and always upsert to nodes table.

File: core/src/stores/postgres.rs
Comment (spolu) [inline core/src/stores/postgres.rs:135]: This is quite brittle isn't it. Any tweak to the query will break I think we need to have that an argument no?

File: core/src/stores/postgres.rs
Comment (spolu) [inline core/src/stores/postgres.rs:52]: This is also quite brittle as any change to the name in the query will break that logic. 

I thiiiiiink we'd be better off with &str table names but we likely need two given the logic you're introducing?

**Review by spolu — COMMENTED** [review]
The fact that we assume some table shorthand names across the whole file is a bit brittle no?

I think I would add 2 arguments: nodeTableAlias and targetTableAlias typed as &str? (we can't have None anymore right? we always have to specify the two?)

Also this is blocked on a backfill right?

**Comment by spolu** [comment]
Risk should be updated to include: "all query-less retrieval actions (include)"

File: core/src/stores/postgres.rs
Comment (tdraier) [reply core/src/stores/postgres.rs:52]: yes we'll need at least the two table names. At first I actually passed the 3 prefixes - maybe passing the 3 full column names would be more clear.

**Comment by tdraier** [comment]
Backfill is done so this one is ready to be merged. 
We're still upserting parents in the 2 tables, in case something goes wrong and that we need to revert.

File: core/src/stores/postgres.rs
Comment (spolu) [inline core/src/stores/postgres.rs:120]: This is dangerous not passing parents would deactiate parent filtering which would potentially have drastic security consequence. Any chance we can make parents_column non optional?

Comment (spolu) [reply core/src/stores/postgres.rs:120]: Looks like we can make parents and timestamp non optional right?

File: core/src/stores/postgres.rs
Comment (spolu) [inline core/src/stores/postgres.rs:3090]: count are really expensive. We don't need that join do we?

File: core/src/stores/postgres.rs
Comment (spolu) [inline core/src/stores/postgres.rs:1710]: I see why we need that JOIN for the count to work (dsn.parents vs dsd.tags_array and dsd.timestamp)

I'm worried this will be too expensive? Any chance we can test at scale vs current state of affairs?

**Review by spolu — COMMENTED** [review]
Left another batch of comments. Much cleaner!!

File: core/src/stores/postgres.rs
Comment (tdraier) [reply core/src/stores/postgres.rs:120]: Yes parents and timestamp are always supposed to be set, so i can get rid of the optional

File: core/src/stores/postgres.rs
Comment (tdraier) [reply core/src/stores/postgres.rs:3086]: It's needed for constraints on `dsn.parents`

**Review by spolu — APPROVED** [review]
LGTM

Very cool. Requires extreme testing of course, in particular in the context of enforcing `view_filter`.

File: core/src/stores/postgres.rs
Comment (spolu) [reply core/src/stores/postgres.rs:1706]: Thanks for the detailed analysis. Let's keep a close eye on it :+1:

File: core/src/stores/postgres.rs
Comment (spolu) [inline core/src/stores/postgres.rs:1897]: do we have guarantee we were always in the case `should_upsert_node = true` since migration?

File: core/src/stores/postgres.rs
Comment (tdraier) [reply core/src/stores/postgres.rs:1706]: I also think we could make the count result optional, as it's just ignore in many queries ( for example the `retrieve_chunks_without_query` just ignores it )

File: core/src/stores/postgres.rs
Comment (tdraier) [reply core/src/stores/postgres.rs:1897]: We were skipping it before to avoid having default values on things upserted by connectors. Now connectors send the correct values, and the backfill has been run. Since all select will start  relying on data_sources_nodes we must ensure nodes are always inserted , even if mimetype/title are not passed. We may still have this ( so cases where we had should_upsert_node = false) as it's not mandatory in public api, but it will now insert with default mimetpye and title.

Comment (spolu) [reply core/src/stores/postgres.rs:1897]: ack

**Pull request merged**
